requests-for-comments/2022-06-01-session-analysis.md

```diff
@@ -0,0 +1,118 @@
+
+![image](https://user-images.githubusercontent.com/4813045/171521347-8f3fd16c-28ab-49ac-9baa-c02097d4c42c.png)
+
+Details: The “auto-bucket” functionality would require more work to figure out exactly how this should work + what the UX should be.
+
+### 5. Count of a specific event per session over time
+
+Proposal: Users accomplish this using formulas. They would make a series that is (a) Session count and (b) total count, and then divide (b) by (a).
+
+![image](https://user-images.githubusercontent.com/4813045/171521365-2b353bb0-7634-45fa-84e3-1463eb351fc0.png)
+
+### 6. Limit a funnel conversion to a single session
+
+Proposal: In funnels, we add “Unique sessions” as an aggregation option. Setting this value would enforce that the sequence of events must occur within a specific session.
+
+![image](https://user-images.githubusercontent.com/4813045/171521401-3b7a234c-24b4-45cd-a3d0-15ab4243efb0.png)
+
+
+### 7. Limit a path conversion to a single session
+
+Proposal: I’d consider this to be a lower priority, but we should update the logic in paths to respect the concept of a session. Today, it uses a 30 min gap between events to split paths - which is essentially the same as sessions, but subtly different.
+
+One issue that we will have to figure out is being backward compatible to data that was generated prior to session support.
+
+## Work breakout:
+
+To accomplish the above, we would need to do the following work:
+
+- Trends:
+    - Add “unique sessions” as an event aggregation option in trends
+    - Add the concept of “session properties”, and specifically the “session length” property to the taxonomic filter.
+    - Allow for session properties to be used as a property aggregation in trends.
+    - Allow for session properties to be used in breakdowns
+    - Add the ability to “auto bucket” when breaking down by numeric properties
+    - Note: while not included in the requirements above, it’d be cool to allow filtering based on session length everywhere. (e.g. total count of pageviews where the corresponding session length > 1 min)
+- Funnels
+    - Add sessions as a aggregation target
+- Paths
+    - Use sessions instead of the 30 min gap (need to consider backward compatibility here)
+
+## Questions + notes
+
+1. Do we need a sessions table?
+    1. If we add session length as a property option, those queries will have to generate a list of session durations on the fly. This query can’t be limited to specific event types because it has to query across all events to ensure it finds the first and last event in the session. As a result it might be a hefty query. To optimize this, we could create a sessions table in clickhouse (probably using AggregatingMergeTree) ****that contains the session_id and it’s length
+    2. My vote is that we don’t over optimize early and try to do without this table at first. We can always add it later.
```

Comment:
I think this would get tricky with backwards compatibility 😬

Reply:
I'd call it a completely new feature? 🤔 - do we currently have sessions exposed in app?

Reply:
I suppose not although just feelsbad if the graph is empty before a certain time

Reply:
This seems really similar to a `first_seen_at` and `last_seen_at` property. I think several users have [asked for this](https://github.com/PostHog/posthog/issues/9683) on persons (and probably other groups). If we were to implement it, I think you're right that it would have to be special cased - because it doesn't make sense with person-on-events. A few other considerations are probably:
* How would we do a backfill on it?
* How would we handle out-of-order events or (event backfills from 3rd parties as a prime example). I think we would just do a max/min function for this one.

I'd be very curious to get the thoughts of someone from the ingestion team on this one. @tiina303 or @yakkomajuri

Reply:
Let's implement these initially as 'computed' properties rather than a separate denormalization initially. If performance is bad we can change ingestion to suit this, but queries are 100x easier to iterate than ingestion.